### ISSUE TITLE
Parfor lowering error

### ISSUE THREAD

**Issue by ehsantn** [IntelLabs/numba#10]
The code below throws an error in parfor lowering (works with sequential lowering). Seems like a jump target is wrong.
```python
@numba.njit(parallel=True)
def logistic_regression(iterations):
    N = 10**7
    D = 10
    g = 2*np.random.ranf(D)-1
    X = 2*np.random.ranf((N,D))-1
    Y = (np.dot(X,g)>0.0)==(np.random.ranf(N)>.80)

    w = 2*np.random.ranf(D)-1
    for i in range(iterations):
        w -= np.dot(((1.0 / (1.0 + np.exp(-Y * np.dot(X,w))) - 1.0) * Y), X)
        R = np.dot(X,w)>0.0
        accuracy = np.sum(R==Y)/N

w = logistic_regression(20)
```

**Comment by ninegua** [comment]
If you dump the parfor before the error happens, you can see that the incoming parfor is wrong:

```
---begin parfor 4---
index_var =  parfor_index.34
LoopNest(index_variable=parfor_index.34, range=0,N,1 correlation=6)
init block:
    $np_g_var.62 = global(np: <module 'numpy' from '/nix/store/g44lywai715lmksw88yia985wswkl5fx-python3.6-numpy-1.12.1/lib/python3.6/site-packages/numpy/__init__.py'>) ['$np_g_var.62']
    $empty_attr_attr.63 = getattr(value=$np_g_var.62, attr=empty) ['$empty_attr_attr.63', '$np_g_var.62']
    $np_typ_var.64 = getattr(value=$np_g_var.62, attr=bool) ['$np_g_var.62', '$np_typ_var.64']
    del $np_g_var.62                         []
    $0.37 = call $empty_attr_attr.63($const0.1, $np_typ_var.64, func=$empty_attr_attr.63, args=[Var($const0.1, test.py (6)), Var($np_typ_var.64, test.py (10))], kws=(), vararg=None) ['$0.37', '$const0.1', '$empty_attr_attr.63', '$np_typ_var.64']
    del $np_typ_var.64                       []
    del $empty_attr_attr.63                  []
label 213:
    $const.40 = const(int, 0)                ['$const.40']
    $sum_var.41 = $const.40                  ['$const.40', '$sum_var.41']
    del $const.40                            []
    $range_g_var.42 = global(range: <class 'range'>) ['$range_g_var.42']
    $range_c_var.43 = call $range_g_var.42(_const0_2, func=$range_g_var.42, args=[Var(_const0_2, test.py (7))], kws=(), vararg=None) ['$range_c_var.43', '$range_g_var.42', '_const0_2']
    del $range_g_var.42                      []
    $iter_var.44 = getiter(value=$range_c_var.43) ['$iter_var.44', '$range_c_var.43']
    del $range_c_var.43                      []
    jump 5                                   []
label 214:
    $iternext_var.46 = iternext(value=$iter_var.44) ['$iter_var.44', '$iternext_var.46']
    $pair_first_var.47 = pair_first(value=$iternext_var.46) ['$iternext_var.46', '$pair_first_var.47']
    $pair_second_var.48 = pair_second(value=$iternext_var.46) ['$iternext_var.46', '$pair_second_var.48']
    del $iternext_var.46                     []
    branch $pair_second_var.48, 6, 7         ['$pair_second_var.48']
label 215:
    del $pair_second_var.48                  []
    $tuple_var.51 = build_tuple(items=[Var($24.3.276, <string> (2)), Var($pair_first_var.47, test.py (10))]) ['$24.3.276', '$pair_first_var.47', '$tuple_var.51']
    $X_val.52 = getitem(value=_0_22, index=$tuple_var.51) ['$X_val.52', '$tuple_var.51', '_0_22']
    del $tuple_var.51                        []
    $g_val.53 = getitem(value=_0_11, index=$pair_first_var.47) ['$g_val.53', '$pair_first_var.47', '_0_11']
    del $pair_first_var.47                   []
    $add_var.54 = $X_val.52 * $g_val.53      ['$X_val.52', '$add_var.54', '$g_val.53']
    del $g_val.53                            []
    del $X_val.52                            []
    $acc_var.55 = inplace_binop(fn=+=, immutable_fn=+, lhs=$sum_var.41, rhs=$add_var.54, static_lhs=<object object at 0x7f8a1a392660>, static_rhs=<object object at 0x7f8a1a392660>) ['$acc_var.55', '$add_var.54', '$sum_var.41']
    del $add_var.54                          []
    $sum_var.41 = $acc_var.55                ['$acc_var.55', '$sum_var.41']
    del $acc_var.55                          []
    jump 5                                   []
...
```

where label 5 is not locally defined.

**Comment by ninegua** [comment]
Somehow the code in get_parfor_params goes back to look at the entire `func_ir`, where part of which have already been lowered. The parfor lowering is destructive, and replaces labels and stuff in-place, so parfors that are already lowered would no longer have an IR that is considered valid in the context of the original function. I think this is where the problem is.

**Comment by ehsantn** [comment]
Ok, this makes sense. Since `get_parfor_params` needs the whole IR, maybe it should be called before lowering and parameters should be saved in parfors themselves.

**Comment by ninegua** [comment]
Right, I was going to suggest do the computation in `get_parfor_params` as the final thing in this parfor pass, and cache the results so that later calls to `get_parfor_params` would not have to redo the computation all over again. So it is the same thing you had in mind.

**Comment by ehsantn** [comment]
Ok, sounds good.

**Comment by ehsantn** [comment]
fixed with e97ac3d5063fa79dfc2201cd6b4779c2847013fb.
